ejson/packages/ejson/lib/src/decoding.dart

```diff
@@ -0,0 +1,279 @@
+  undefinedOr<T>(f) => f<UndefinedOr<T>>();
+  TypePlus.addFactory(undefinedOr);
+  TypePlus.addFactory(<T>(f) => f<Defined<T>>(), superTypes: [undefinedOr]);
+  TypePlus.addFactory(<T>(f) => f<Undefined<T>>(), superTypes: [undefinedOr]);
+  TypePlus.add<Key>();
+  TypePlus.add<ObjectId>();
+  TypePlus.add<Uuid>();
+
+  return CombinedMapView([customDecoders, commonDecoders]);
+}();
+
+T fromEJson<T>(EJsonValue ejson) {
+  final type = T;
+  final nullable = type.isNullable;
+  final decoder = nullable ? _decodeNullable : decoders[type.base];
+  if (decoder == null) {
+    throw MissingDecoder(ejson, type);
+  }
+  final args = nullable ? [type.nonNull] : type.args;
+  if (args.isEmpty) {
+    return decoder(ejson); // minor optimization
+  }
+  return decoder.callWith(typeArguments: args, parameters: [ejson]);
+}
+
+// Important to return `T` as opposed to [Never] for type inference to work
+T raiseInvalidEJson<T>(Object? value) => throw InvalidEJson<T>(value);
+
+dynamic _decodeAny(EJsonValue ejson) {
+  return switch (ejson) {
+    null => null,
+    bool b => b,
+    double d => d, // relaxed mode
+    int i => i, // relaxed mode
+    String s => s,
+    {'\$date': _} => _decodeDate(ejson),
+    {'\$maxKey': _} => _decodeKey(ejson),
+    {'\$minKey': _} => _decodeKey(ejson),
+    {'\$numberDouble': _} => _decodeDouble(ejson),
+    {'\$numberInt': _} => _decodeInt(ejson),
+    {'\$numberLong': _} => _decodeInt(ejson),
+    {'\$regex': _} => _decodeString(ejson),
+    {'\$symbol': _} => _decodeSymbol(ejson),
+    {'\$undefined': _} => _decodeUndefined(ejson),
+    {'\$oid': _} => _decodeObjectId(ejson),
+    {'\$binary': {'base64': _, 'subType': '04'}} => _decodeUuid(ejson),
+    List _ => _decodeArray(ejson),
+    Map _ => _tryDecodeCustom(ejson) ??
+        _decodeDocument(ejson), // other maps goes last!!
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+dynamic _tryDecodeCustom(json) {
+  for (final decoder in customDecoders.values) {
+    try {
+      return decoder(json);
+    } catch (_) {
+      // ignore
+    }
+  }
+  return null;
+}
+
+List<T> _decodeArray<T>(EJsonValue ejson) {
+  return switch (ejson) {
+    Iterable i => i.map((ejson) => fromEJson<T>(ejson)).toList(),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+bool _decodeBool(EJsonValue ejson) {
+  return switch (ejson) {
+    bool b => b,
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+DateTime _decodeDate(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$date': String s} => DateTime.parse(s), // relaxed mode
+    {'\$date': {'\$numberLong': int i}} =>
+      DateTime.fromMillisecondsSinceEpoch(i),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+Defined<T> _decodeDefined<T>(EJsonValue ejson) {
+  if (ejson case {'\$undefined': 1}) raiseInvalidEJson(ejson);
+  return Defined<T>(fromEJson<T>(ejson));
+}
+
+Map<K, V> _decodeDocument<K, V>(EJsonValue ejson) {
+  return switch (ejson) {
+    Map m => m.map((key, value) => MapEntry(key as K, fromEJson<V>(value))),
```

Comment:
This actually works for K != String, but obviously that will not render to json string, and hence it is not really valid ejson either.

To give an example
```dart
fromEJson<Map<int, bool>>(toEJson({1: true}));
```
actually works, but trying to encode the intermediate as json will obviously fail, since 1 is not a string and hence not a valid key in json.